File: routes/scripts.ts
import express, { Response, Request } from "express";
import database from "../db";
import { ERRORS } from "../errors";
import authorization from "../middleware";
const scripts = express.Router();
scripts.use(authorization);
scripts
  .route("/")
  .get(async (req: Request, res: Response) => {
    const { fields, limit, order_by, order_direction } = req.headers;
    let query;
    if (fields) {
      query = database.from("scripts").select(fields as string);
    } else {
      query = database.from("scripts").select();
    }
    if (limit) {
      query.limit(parseInt(limit as string));
    }
    if (order_by) {
      if (order_direction == "desc") {
        query.order(order_by as string, { ascending: false });
      } else {
        query.order(order_by as string, { ascending: true });
      }
    }
    const { data, error } = await query;
    if (error) {
      res.status(503).send(error);
      return;
    }
    if (data.length == 0) {
      res.status(404).send(ERRORS.NO_FILTERED_RESULTS);
      return;
    }
    res.status(200).send(data);
  })
  .post((req: Request, res: Response) => {
    const route = req.baseUrl + req.path;
    const method = req.method;
    res.status(501).send({ route, method, ...ERRORS.UNIMPLEMENTED });
  });

scripts.route("/filter").get(async (req: Request, res: Response) => {
  const {
    max_year,
    min_year,
    living,
    direction,
    continent,
    fields,
    limit,
    order_by,
    order_direction,
  } = req.headers;
  let query;
  if (fields) {
    query = database.from("scripts").select(fields as string);
  } else {
    query = database.from("scripts").select();
  }
  if (max_year) {
    query = query.lte("year", parseInt(max_year as string));
  }
  if (min_year) {
    query = query.gte("year", parseInt(min_year as string));
  }
  if (living) {
    query = query.eq("living", JSON.parse(living as string));
  }
  if (direction) {
    query = query.eq("direction", direction);
  }
  if (continent) {
    query = query.contains("continents", [continent]);
  }
  if (limit) {
    query = query.limit(parseInt(limit as string));
  }
  if (order_by) {
    if (order_direction == "desc") {
      query.order(order_by as string, { ascending: false });
    } else {
      query.order(order_by as string, { ascending: true });
    }
  }
  try {
    const { data, error } = await query;
    if (error) {
      res.status(503).send(error);
      return;
    }
    if (data.length == 0) {
      res.status(404).send(ERRORS.NO_FILTERED_RESULTS);
      return;
    }
    res.status(200).send(data);
  } catch (err) {
    res.status(500).send(err);
  }
});

scripts
  .route("/:id")
  .get(async (req: Request, res: Response) => {
    const id = parseInt(req.params.id);
    const { fields } = req.headers;
    let query;
    if (fields) {
      query = database.from("scripts").select(fields as string);
    } else {
      query = database.from("scripts").select();
    }
    try {
      const { data, error } = await query.eq("id", id);
      if (error) {
        res.status(503).send(error);
        return;
      }
      if (data.length == 0) {
        res.status(404).send(ERRORS.ID_NOT_FOUND);
        return;
      }
      res.status(200).send(data);
    } catch (err) {
      res.status(500).send(err);
    }
  })
  .put((req: Request, res: Response) => {
    const route = req.baseUrl + req.path;
    const method = req.method;
    res.status(501).send({ route, method, ...ERRORS.UNIMPLEMENTED });
  });

export default scripts;

File: handler/scripts.go
package handler

import (
	"script_sweep_go/database"
	m "script_sweep_go/models"
	r "script_sweep_go/responses"
	"strconv"

	"github.com/gofiber/fiber/v2"
	"gorm.io/gorm"
)

func AllScripts(c *fiber.Ctx) error {
	filters := new(m.ScriptFilters)
	err := c.ReqHeaderParser(filters)
	if err != nil {
		return c.Status(400).JSON(r.BadRequestHeader(err))
	}
	db := database.DB
	var rows []m.ScriptRow
	fields := []string{"id", "name", "living", "ranges", "year", "direction", "continents", "link"}
	if len(filters.Fields) > 0 {
		fields = filters.Fields
	}
	// possibly change this to optimize it
	var query *gorm.DB
	if filters.Limit != "" {
		limit, _ := strconv.Atoi(filters.Limit)
		query = db.Table("scripts").Select(fields).Limit(limit)
	} else {
		query = db.Table("scripts").Select(fields)
	}
	if filters.MaxYear != "" {
		query = query.Where("year < ?", filters.MaxYear)
	}
	if filters.MinYear != "" {
		query = query.Where("year > ?", filters.MinYear)
	}
	if filters.Living != "" {
		query = query.Where("living = ?", filters.Living)
	}
	if filters.Continent != "" {
		query = query.Where("? = ANY (continents)", []string{filters.Continent})
	}
	if filters.Direction != "" {
		query = query.Where("direction = ?", filters.Direction)
	}
	if filters.OrderBy != "" {
		if filters.OrderDirection != "" {
			query = query.Order(filters.OrderBy + " DESC")
		} else {
			query = query.Order(filters.OrderBy)
		}
	}
	query.Find(&rows)
	filtered_response, err := m.FilterFields(rows, fields)
	if err != nil {
		return c.Status(503).JSON(r.BadRequestFields(err))
	}
	if len(filtered_response) <= 0 {
		return c.Status(404).JSON(r.TooManyFilters())
	}
	return c.Status(200).JSON(r.AllScriptsRes(filtered_response))
}

func ScriptByID(c *fiber.Ctx) error {
	id := c.Params("id")
	int_id, _ := strconv.Atoi(id)
	filters := new(m.ScriptFilters)
	err := c.ReqHeaderParser(filters)
	if err != nil {
		return c.Status(400).JSON(r.BadRequestHeader(err))
	}
	db := database.DB
	var rows []m.ScriptRow
	// possibly change this to optimize it
	fields := []string{"id", "name", "living", "ranges", "year", "direction", "continents", "link"}
	if len(filters.Fields) > 0 {
		fields = filters.Fields
	}
	db.Table("scripts").Where("id = ?", int_id).Find(&rows)
	// possibly change this to optimize it
	filtered_response, err := m.FilterFields(rows, fields)
	if err != nil {
		panic(err)
	}
	return c.Status(200).JSON(r.SingleScriptRes(filtered_response, id))
}
